data/transactions/logic/assembler.go

```diff
@@ -1189,6 +1187,197 @@ func (ops *OpStream) checkArgs(spec OpSpec) {
+				ops.tpusha(spec.Returns)
+			}
+			j = j + spec.Details.Size
+		}
+		//Not sure if slices require you to make copies like this
+		out = make([]StackType, len(ops.typeStack))
+		copy(out, ops.typeStack)
+		blocks[i].stackDelta = out
+		blocks[i].expectedStackTop = top
+
+	}
+}
+func (ops *OpStream) addBlockJumps(indexToBlock map[int]int) {
+	blocks := ops.blocks
+	raw := ops.pending.Bytes()
+	var spec OpSpec
+	for i := range blocks {
+		spec = opsByOpcode[ops.Version][raw[blocks[i].endIndex]]
+		switch spec.Name {
+		case "bz", "bnz", "b":
+			x, ok := indexToBlock[blocks[i].jumpTo]
+			if ok {
+				blocks[i].jumpTo = x
+			} else {
+				blocks[i].jumpTo = exiting //if a label index is not in indexToBlock it is an end label or super weird
+			}
+		}
+	}
+}
+
+//createBlocks sets up the basic blocks to be used in static analysis; note no block is added for labels at end of program
+func (ops *OpStream) createBlocks(reverseLabels map[int]string) map[int]int {
+	indexToBlock := make(map[int]int)
+	var blocks []BasicBlock
+	raw := ops.pending.Bytes()
+	startI := 0
+	i := 0         //index over raw
+	refsIndex := 0 //taking advantage of fact labelRefs is "sorted" automatically
+	var spec OpSpec
+	var _ string
+	var ok bool
+	for i < len(raw) {
+		spec = opsByOpcode[ops.Version][raw[i]]
+		_, ok = reverseLabels[i+spec.Details.Size] //checks if next instruction was a label
+		if i == ops.labelReferences[refsIndex].position {
+			//We put the start indices of where we jump to b/c we will have map from start indices to blocks after
+			if spec.Name == "bz" || spec.Name == "bnz" {
+				blocks = append(blocks, BasicBlock{startIndex: startI, endIndex: i, jumpTo: ops.labels[ops.labelReferences[refsIndex].label], flowTo: len(blocks) + 1})
+				indexToBlock[startI] = len(blocks) - 1
+			} else if spec.Name == "b" {
+				blocks = append(blocks, BasicBlock{startIndex: startI, endIndex: i, jumpTo: ops.labels[ops.labelReferences[refsIndex].label], flowTo: nowhere})
+				indexToBlock[startI] = len(blocks) - 1
+			}
+			refsIndex++
+			i = i + spec.Details.Size
+			startI = i
+			//eventually need to deal with callsub
+		} else if spec.Name == "return" || spec.Name == "err" {
+			blocks = append(blocks, BasicBlock{startIndex: startI, endIndex: i, jumpTo: exiting, flowTo: nowhere})
+			indexToBlock[startI] = len(blocks) - 1
+			i = i + spec.Details.Size
+			startI = i
+		} else if spec.Name == "assert" {
+			blocks = append(blocks, BasicBlock{startIndex: startI, endIndex: i, jumpTo: exiting, flowTo: len(blocks) + 1})
+			indexToBlock[startI] = len(blocks) - 1
+			i = i + spec.Details.Size
+			startI = i
+		} else if ok {
+			if startI != i {
+				//Want to make sure we are not duplicating blocks if there is a branch right before a label
+				blocks = append(blocks, BasicBlock{startIndex: startI, endIndex: i, jumpTo: nowhere, flowTo: len(blocks) + 1})
+				indexToBlock[startI] = len(blocks) - 1
+				startI = i
+			}
+			i = i + spec.Details.Size
+		}
```

Comment:
It seems you are not incrementing `i` unless one of these clauses is true, which tells me you probably always expect that? If so, how about an error if it's not true?

Also, maybe use the three clause version of `for` to initialize `i`, perform the check you're doing now, and advance `i` by Details.size in one place.